dm/worker/task_checker.go

```diff
@@ -0,0 +1,306 @@
+func NewRealTaskStatusChecker(cfg CheckerConfig, w *Worker) TaskStatusChecker {
+	tsc := &realTaskStatusChecker{
+		cfg: cfg,
+		l:   log.With(zap.String("component", "task checker")),
+		w:   w,
+		bc:  newBackoffController(),
+	}
+	tsc.closed.Set(closedTrue)
+	return tsc
+}
+
+// Init implements TaskStatusChecker.Init
+func (tsc *realTaskStatusChecker) Init() error {
+	// just check configuration of backoff here, lazy creates backoff counter,
+	// as we can't get task information before dm-worker starts
+	_, err := backoff.NewBackoff(tsc.cfg.BackoffFactor, tsc.cfg.BackoffJitter, tsc.cfg.BackoffMin, tsc.cfg.BackoffMax)
+	return terror.WithClass(err, terror.ClassDMWorker)
+}
+
+// Start implements TaskStatusChecker.Start
+func (tsc *realTaskStatusChecker) Start() {
+	tsc.wg.Add(1)
+	go func() {
+		defer tsc.wg.Done()
+		tsc.run()
+	}()
+}
+
+// Close implements TaskStatusChecker.Close
+func (tsc *realTaskStatusChecker) Close() {
+	if !tsc.closed.CompareAndSwap(closedFalse, closedTrue) {
+		return
+	}
+
+	if tsc.cancel != nil {
+		tsc.cancel()
+	}
+	tsc.wg.Wait()
+}
+
+func (tsc *realTaskStatusChecker) run() {
+	tsc.ctx, tsc.cancel = context.WithCancel(context.Background())
+	tsc.closed.Set(closedFalse)
+	ticker := time.NewTicker(tsc.cfg.CheckInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-tsc.ctx.Done():
+			tsc.l.Info("worker task checker exits")
+			return
+		case <-ticker.C:
+			tsc.check()
+		}
+	}
+}
+
+// isResumableError checks the error message and returns whether we need to
+// resume the task and retry
+func isResumableError(err *pb.ProcessError) bool {
+	// not elegant code, because TiDB doesn't expose some error
+	unsupportedDDLMsgs := []string{
+		"can't drop column with index",
+		"unsupported add column",
+		"unsupported modify column",
+		"unsupported modify",
+		"unsupported drop integer primary key",
+	}
+	if err.Type == pb.ErrorType_ExecSQL {
+		for _, msg := range unsupportedDDLMsgs {
+			if strings.Contains(err.Msg, msg) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (tsc *realTaskStatusChecker) getResumeStrategy(stStatus *pb.SubTaskStatus, duration time.Duration) ResumeStrategy {
+	// task that is not paused or paused manually, just ignore it
+	if stStatus == nil || stStatus.Stage != pb.Stage_Paused || stStatus.Result == nil || stStatus.Result.IsCanceled {
+		return ResumeIgnore
+	}
+
+	// TODO: use different strategies based on the error detail
+	for _, processErr := range stStatus.Result.Errors {
+		if !isResumableError(processErr) {
+			return ResumeNoSense
+		}
+	}
+
+	// auto resume interval does not exceed backoff duration, skip this paused task
+	if time.Since(tsc.bc.latestResumeTime[stStatus.Name]) < duration {
+		return ResumeSkip
+	}
+
+	return ResumeDispatch
+}
+
+func (tsc *realTaskStatusChecker) check() {
+	allSubTaskStatus := tsc.w.getAllSubTaskStatus()
+	tasks := make(map[string]struct{}, len(allSubTaskStatus))
```

Comment:
can `getAllSubTaskStatus` return a `map` instead a `slice` (taskName -> status)? if so, this `tasks` is useless.